File: C++/stacks_and_queues/sort_stack.cpp
#include <iostream>
#include <stack>
using namespace std;

void sortStack(stack<int> *st){
    stack<int> sorted; // Create an empty stack called sorted to store the sorted elements.

    // While the input stack is not empty, do the following:
    while(!(*st).empty()){
        int tmp = (*st).top(); // Pop the top element from the input stack and store it in a temporary variable called tmp.
        (*st).pop();

        // While the sorted stack is not empty and the top element of the sorted stack is greater than tmp, do the following:
        while(!sorted.empty() && tmp < sorted.top()){
            (*st).push(sorted.top()); // Pop the top element from the sorted stack and push it back onto the input stack.
            sorted.pop();
        }

        sorted.push(tmp); // Push tmp onto the sorted stack.
    }

    // Once the input stack is empty, sorted stack will contain the elements in ascending order. 
    // To reverse the order and obtain the sorted elements in the input stack, do the following:
    while(!sorted.empty()){
        (*st).push(sorted.top()); // Pop the top element from the sorted stack and push it back onto the input stack.
        sorted.pop();
    }

    // The input stack now contains the sorted elements in ascending order.
}


int main(){
stack<int> mystack; // Create an empty stack of integers called mystack.

int arr[] = {6, 4, 8, 9, 10, 99, 7, 1, 100}; // Create an array of integers called arr.

// Loop through each element of the array and push it onto the stack.
for(int i=0; i<sizeof(arr)/sizeof(arr[0]); i++){
    cout<<"Push in element "<<arr[i]<<endl; // Print a message to indicate which element is being pushed onto the stack.
    mystack.push(arr[i]); // Push the current element onto the stack.
}

sortStack(&mystack); // Sort the stack using the sortStack function.

cout<<"The sorted stack is: "<<endl; // Print a message to indicate that the sorted stack is being printed.

// Loop through each element of the sorted stack and print it out.
while(!mystack.empty()){
    cout<<mystack.top()<<'\n'; // Print the top element of the stack.
    mystack.pop(); // Pop the top element off the stack.
}

}

File: Python/stacks_and_queues/sort_stack.py
# 3.5 Sort Stacks
import unittest
from stack import Stack

class SortedStack(Stack):
    """
    A stack data structure that maintains its items in sorted order, with the smallest item at the top of the stack.
    """

    def __init__(self):
        """
        Initializes an empty SortedStack object and creates a temporary stack object to help with sorting.
        """
        super().__init__()
        self.temp_stack = Stack()

    def push(self, item):
        """
        Pushes an item onto the stack while maintaining the sorted order.
        
        If the new item is smaller than or equal to the current top item of the stack, it is simply pushed onto the stack.
        If the new item is larger than the current top item, it is pushed onto the stack in sorted order.
        
        Args:
            item: The item to be pushed onto the stack.
        """
        if self.is_empty() or item < self.peek():
            super().push(item)
        else:
            # Move items from the original stack to the temporary stack until a larger item than the new item is found.
            while self.peek() is not None and item > self.peek():
                self.temp_stack.push(self.pop())
                
            # Push the new item onto the original stack in sorted order.
            super().push(item)
            
            # Move the items from the temporary stack back onto the original stack in their original order.
            while not self.temp_stack.is_empty():
                super().push(self.temp_stack.pop())



class Tests(unittest.TestCase):
    """
    A collection of unit tests for the SortedStack class.
    """

    def test_push_one(self):
        """
        Tests pushing a single item onto the stack.
        """
        queue = SortedStack()
        queue.push(1)
        assert len(queue) == 1

    def test_push_two(self):
        """
        Tests pushing two items onto the stack.
        """
        queue = SortedStack()
        queue.push(1)
        queue.push(2)
        assert len(queue) == 2

    def test_push_three(self):
        """
        Tests pushing three items onto the stack.
        """
        queue = SortedStack()
        queue.push(1)
        queue.push(2)
        queue.push(3)
        assert len(queue) == 3

    def test_pop_one(self):
        """
        Tests popping a single item from the stack.
        """
        queue = SortedStack()
        queue.push(1)
        assert queue.pop() == 1

    def test_pop_two(self):
        """
        Tests popping two items from the stack.
        """
        queue = SortedStack()
        queue.push(1)
        queue.push(2)
        assert queue.pop() == 1
        assert queue.pop() == 2

    def test_pop_three(self):
        """
        Tests popping three items from the stack.
        """
        queue = SortedStack()
        queue.push(1)
        queue.push(2)
        queue.push(3)
        assert queue.pop() == 1
        assert queue.pop() == 2
        assert queue.pop() == 3

    def test_push_mixed(self):
        """
        Tests pushing a mix of sorted and unsorted items onto the stack.
        """
        queue = SortedStack()
        queue.push(3)
        queue.push(2)
        queue.push(1)
        queue.push(4)
        assert queue.pop() == 1
        assert queue.pop() == 2
        assert queue.pop() == 3
        assert queue.pop() == 4
        
if __name__ == '__main__':
    # Create a TestSuite object and add all the tests from the Tests class
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(Tests))

    # Run the tests and print the results
    runner = unittest.TextTestRunner()
    result = runner.run(suite)
    print(result)
